### PR TITLE
[testing] Make tests great again

### PR DESCRIPTION
**Pull request by RomanenkoDenys**
## Description

For all controllers we use special helper to construct image name with tag for container.
Fixed local running of matrix tests.
## Why do we need it, and what problem does it solve?

We want to get the same results when we run tests locally and from CI.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests are passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: tests
type: chore
summary: fixed local running of matrix tests.
impact: all controllers should restart
impact_level: high
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

File: testing/matrix/linter/rules/rules.go [hunk 1]
```diff
@@ -177,13 +197,35 @@ func containerImageTagLatest(object storage.StoreObject, containers []v1.Contain
...
+			if !matched {
+				return errors.NewLintRuleError("CONTAINER003",
+					object.Identity()+"; container = "+c.Name,
+					nil,
+					"Image tag should be in form sha256 hash",
+				)
+			}
+		default:
+			if imageParts[1] != "imageHash" {
+				return errors.NewLintRuleError("CONTAINER003",
```
Comment (diafour) [inline testing/matrix/linter/rules/rules.go:222]: I propose make unique rule identifiers to search in code in case of validation errors.

```suggestion
				return errors.NewLintRuleError("CONTAINER004",
```

File: .github/ci_templates/tests.yml [hunk 2]
```diff
@@ -7,8 +7,8 @@ docker_options: '-w /deckhouse -v ${{env.HOME}}/go-pkg-cache:/go/pkg'
 
 {!{ define "matrix_run_args" }!}
 # <template: matrix_run_args>
-args: 'go test -timeout=${{env.TEST_TIMEOUT}} ./testing/matrix/ -v'
-docker_options: '-w /deckhouse -v ${{env.HOME}}/go-pkg-cache:/go/pkg'
+args: 'sh -c "rm -f ee/modules/030-cloud-provider-openstack/candi ee/modules/030-cloud-provider-vsphere/candi; go test -timeout=${{env.TEST_TIMEOUT}} ./testing/matrix/ -v ; ln -s /deckhouse/candi/cloud-providers/openstack/ ee/modules/030-cloud-provider-openstack/candi ; ln -s /deckhouse/candi/cloud-providers/vsphere/ ee/modules/030-cloud-provider-vsphere/candi"'
```
Comment (diafour) [inline .github/ci_templates/tests.yml:10]: Restoring links with `ln -s` is not needed, each job has its own copy of repository.

But what is the problem with `ee/modules/*/candi`?

File: .github/ci_templates/tests.yml [hunk 3]
```diff
@@ -75,3 +75,29 @@ steps:
       docker run {!{ $run.docker_options }!} ${TESTS_IMAGE_NAME} {!{ $run.args }!}
 # </template: tests_template>
 {!{- end -}!}
+
+{!{ define "tests_before_build_template" }!}
```
Comment (diafour) [inline .github/ci_templates/tests.yml:79]:
```suggestion
{!{/* TODO change args in tests_template and merge two templates */}!}
{!{ define "tests_before_build_template" }!}
```

File: .github/workflow_templates/build-and-test_release.yml [hunk 4]
```diff
@@ -121,7 +121,7 @@ jobs:
     needs:
       - git_info
       - build_fe
-{!{ tmpl.Exec "tests_template" (slice $ctx "matrix" "build_fe") | strings.Indent 4 }!}
+{!{ tmpl.Exec "tests_before_build_template" (slice $ctx "matrix") | strings.Indent 4 }!}
```
Comment (diafour) [inline .github/workflow_templates/build-and-test_release.yml:124]: It seems build_fe is redundant now. (Github stops me to suggest here)

File: ee/fe/modules/500-basic-auth/templates/deployment.yaml [hunk 5]
```diff
@@ -44,7 +44,7 @@ spec:
       containers:
       - name: nginx
         {{- include "helm_lib_module_container_security_context_read_only_root_filesystem" . | nindent 8 }}
-        image: "{{ $.Values.global.modulesImages.registry }}:{{ $.Values.global.modulesImages.tags.basicAuth.nginx }}"
+        image: {{ include "helm_lib_module_image" (list $ "nginx") }}
```
Comment (apolovov) [inline ee/fe/modules/500-basic-auth/templates/deployment.yaml:47]: I'll be surprised if there won't be problems with quotations. Due to ':' symbol in image string.

File: ee/modules/600-flant-integration/templates/madison/deployment.yaml [hunk 6]
```diff
@@ -56,7 +56,7 @@ spec:
       containers:
       - name: nginx
         {{- include "helm_lib_module_container_security_context_read_only_root_filesystem" $ | nindent 8 }}
-        image: "{{ $.Values.global.modulesImages.registry }}:{{ $.Values.global.modulesImages.tags.flantIntegration.madisonProxy }}"
+        image: {{ include "helm_lib_module_image" (list $ "madisonProxy") }}
```
Comment (shvgn) [inline ee/modules/600-flant-integration/templates/madison/deployment.yaml:59]: Wrapping quotes let auto-format manifests simetimes :)

**Review by shvgn — DISMISSED** [review]

File: testing/matrix/linter/rules/rules.go [hunk 7]
```diff
@@ -177,13 +197,35 @@ func containerImageTagLatest(object storage.StoreObject, containers []v1.Contain
...
+		if imageParts[0] != defaultRegistry {
 			return errors.NewLintRuleError("CONTAINER003",
 				object.Identity()+"; container = "+c.Name,
 				nil,
-				"Image tag \"latest\" used",
+				"Image name should be "+defaultRegistry,
 			)
 		}
+
+		r := regexp.MustCompile("[A-Fa-f0-9]{64}")
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:209]: Compiling regex in a loop is not efficient. Consider moving it out of the loop/function.

File: testing/matrix/linter/rules/modules/images.go [hunk 8]
```diff
@@ -33,7 +33,8 @@ func skipModuleImageNameIfNeeded(filePath string) bool {
 		// Following images will be removed soon
 		"/deckhouse/modules/110-istio/images/operator-v1x9x1trustca/Dockerfile",
 		"/deckhouse/modules/110-istio/images/pilot-v1x9x1trustca/Dockerfile",
-		"/deckhouse/modules/110-istio/images/proxyv2-v1x9x1trustca/Dockerfile":
+		"/deckhouse/modules/110-istio/images/proxyv2-v1x9x1trustca/Dockerfile",
```
Comment (nabokihms) [inline testing/matrix/linter/rules/modules/images.go:36]: I suppose we do not need images bove anymore.

File: testing/matrix/linter/rules/rules.go [hunk 9]
```diff
@@ -166,8 +172,22 @@ func containerEnvVariablesDuplicates(object storage.StoreObject, containers []v1
 	return errors.EmptyRuleError
 }
 
-func containerImageTagLatest(object storage.StoreObject, containers []v1.Container) errors.LintRuleError {
+func shouldSkipModuleContainer(module string, container string) bool {
+	if module == "okmeter" && container == "okagent" {
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:176]: It would be best to add comments on why these images are ignored.

File: testing/matrix/linter/rules/rules.go [hunk 10]
```diff
@@ -30,6 +33,8 @@ import (
 	"github.com/deckhouse/deckhouse/testing/matrix/linter/utils"
 )
 
+const defaultRegistry = "registry.deckhouse.io/deckhouse/fe"
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:36]: This one should not be hardcoded in tests. If some test string is required, it is better to use something abstract.

File: testing/matrix/linter/rules/rules.go [hunk 11]
```diff
@@ -177,13 +197,35 @@ func containerImageTagLatest(object storage.StoreObject, containers []v1.Contain
...
 			)
 		}
-		if imageParts[1] == "latest" {
+
+		if imageParts[0] != defaultRegistry {
 			return errors.NewLintRuleError("CONTAINER003",
 				object.Identity()+"; container = "+c.Name,
 				nil,
-				"Image tag \"latest\" used",
+				"Image name should be "+defaultRegistry,
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:205]: Not `should be`. You meant something like `All images must be deployed from the same default registry - defaultRegistry`.

File: testing/matrix/linter/rules/rules.go [hunk 12]
```diff
@@ -166,8 +172,22 @@ func containerEnvVariablesDuplicates(object storage.StoreObject, containers []v1
...
+	return false
+}
+
+func containerImageTagCheck(object storage.StoreObject, containers []v1.Container) errors.LintRuleError {
 	for _, c := range containers {
+		if shouldSkipModuleContainer(object.Unstructured.GetName(), c.Name) {
+			continue
+		}
+
 		imageParts := strings.Split(c.Image, ":")
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:191]: I'd like to avoid splitting and suggest using something like this https://github.com/google/go-containerregistry/blob/45aaa6c000ebf2acacad9beef7c82667ec51662c/pkg/name/ref.go#L39-L49

File: testing/matrix/linter/rules/rules.go [hunk 13]
```diff
@@ -177,13 +197,35 @@ func containerImageTagLatest(object storage.StoreObject, containers []v1.Contain
...
+
+		r := regexp.MustCompile("[A-Fa-f0-9]{64}")
+		switch os.Getenv("CI") {
+		case "true":
+			matched := r.MatchString(imageParts[1])
+			if !matched {
+				return errors.NewLintRuleError("CONTAINER003",
+					object.Identity()+"; container = "+c.Name,
+					nil,
+					"Image tag should be in form sha256 hash",
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:217]: Hmm. I doubt that `[A-Fa-f0-9]{64}` is a valid regex for sha256.

File: testing/matrix/linter/rules/rules.go [hunk 14]
```diff
@@ -177,13 +197,35 @@ func containerImageTagLatest(object storage.StoreObject, containers []v1.Contain
...
 				nil,
-				"Image tag \"latest\" used",
+				"Image name should be "+defaultRegistry,
 			)
 		}
+
+		r := regexp.MustCompile("[A-Fa-f0-9]{64}")
+		switch os.Getenv("CI") {
+		case "true":
+			matched := r.MatchString(imageParts[1])
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:212]: `MatchString` returns `true` for all strings with length >= 64. It would be best if you wrapped your regexp in `$^` to make it a full match.

File: testing/matrix/linter/rules/rules.go [hunk 15]
```diff
@@ -85,6 +90,7 @@ type ObjectLinter struct {
 	Module         utils.Module
 	Values         string
 	EnabledModules map[string]struct{}
+	HelmTemplates  []*chart.File
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:93]: This property is not used anywhere :thinking:

File: ee/fe/modules/500-basic-auth/templates/deployment.yaml (hunk 5 above)
Comment (nabokihms) [reply ee/fe/modules/500-basic-auth/templates/deployment.yaml:47]: It works when the colon symbol is in the middle of a string, e.g., `a:b`.

File: .github/workflows/build-and-test_release.yml [hunk 16]
```diff
@@ -1462,23 +1462,15 @@ jobs:
       # </template: login_rw_registry_step>
       - name: Run tests
         env:
-          TESTS_IMAGE_NAME: ${{needs.build_fe.outputs.tests_image_name}}
+          TESTS_IMAGE_NAME: "registry.deckhouse.io/base_images/golang:1.16.3-alpine3.12@sha256:371dc6bf7e0c7ce112a29341b000c40d840aef1dbb4fdcb3ae5c0597e28f3061"
```
Comment (nabokihms) [inline .github/workflows/build-and-test_release.yml:1465]:
```suggestion
          TESTS_IMAGE_NAME: "registry.deckhouse.io/base_images/golang:1.16.3-alpine3.12@sha256:371dc6bf7e0c7ce112a29341b000c40d840aef1dbb4fdcb3ae5c0597e28f3061"
```
I suppose this one should be reverted.

File: .github/workflows/build-and-test_dev.yml [hunk 17]
```diff
@@ -923,23 +924,15 @@ jobs:
...
-          TESTS_IMAGE_NAME=$(echo ${TESTS_IMAGE_NAME} | base64 -d | gunzip)
-
           # Print image name in uppercase to prevent hiding non-secret registry host stored in secret.
           echo "Tests image name: '${TESTS_IMAGE_NAME}'" | tr '[:lower:]' '[:upper:]'
           echo "⚓️ 📥 [$(date -u)] Pull 'tests' image..."
           docker pull ${TESTS_IMAGE_NAME}
           echo "⚓️ 🏎 [$(date -u)] Run tests..."
-          docker run -w /deckhouse -v ${{env.HOME}}/go-pkg-cache:/go/pkg ${TESTS_IMAGE_NAME} go test -timeout=${{env.TEST_TIMEOUT}} ./testing/matrix/ -v
-    # </template: tests_template>
+          docker run -w /deckhouse -v ${{github.workspace}}:/deckhouse -e "TERM=xterm-256color" -e "CGO_ENABLED=0" -v ${{env.HOME}}/go-pkg-cache:/go/pkg ${TESTS_IMAGE_NAME} sh -c "rm -f ee/modules/030-cloud-provider-openstack/candi ee/modules/030-cloud-provider-vsphere/candi; go test -timeout=${{env.TEST_TIMEOUT}} ./testing/matrix/ -v ; ln -s /deckhouse/candi/cloud-providers/openstack/ ee/modules/030-cloud-provider-openstack/candi ; ln -s /deckhouse/candi/cloud-providers/vsphere/ ee/modules/030-cloud-provider-vsphere/candi"
```
Comment (nabokihms) [inline .github/workflows/build-and-test_dev.yml:934]: Can we use make now to run tests the same way locally and on CI?

File: Makefile [hunk 18]
```diff
@@ -81,7 +81,10 @@ tests-modules: ## Run unit tests for modules hooks and templates.
 
 tests-matrix: ## Test how helm templates are rendered with different input values generated from values examples.
   ##~ Options: FOCUS=module-name
+	rm -f ee/modules/030-cloud-provider-openstack/candi ee/modules/030-cloud-provider-vsphere/candi
 	go test ./testing/matrix/ -v
+	ln -s /deckhouse/candi/cloud-providers/openstack/ ee/modules/030-cloud-provider-openstack/candi
```
Comment (nabokihms) [inline Makefile:86]: TODO: Need to find a way to maintain these symlinks.

File: .github/ci_templates/tests.yml (hunk 2 above)
Comment (RomanenkoDenys) [reply .github/ci_templates/tests.yml:10]: When we try to run tests with simlinks, we get an error - 
```
=== RUN   TestMatrix/cloud-provider-openstack
panic on linter run occurred: chart load: error evaluating symlink /deckhouse/ee/modules/030-cloud-provider-openstack/candi: lstat /Users/Ash/work/flant/deckhouse-oss/candi/cloud-providers/openstack: no such file or directory
=== RUN   TestMatrix/cloud-provider-vsphere
panic on linter run occurred: chart load: error evaluating symlink /deckhouse/ee/modules/030-cloud-provider-vsphere/candi: lstat /Users/Ash/work/flant/deckhouse-oss/candi/cloud-providers/vsphere: no such file or directory
```
Helm chart render lib cannot parse inconsistent synlinks.

File: .github/workflow_templates/build-and-test_release.yml (hunk 4 above)
Comment (RomanenkoDenys) [reply .github/workflow_templates/build-and-test_release.yml:124]: tests are run before the ce/ee/fe build, so a third parameter is not needed.

File: ee/fe/modules/500-basic-auth/templates/deployment.yaml (hunk 5 above)
Comment (RomanenkoDenys) [reply ee/fe/modules/500-basic-auth/templates/deployment.yaml:47]: @nabokihms says it right

File: ee/modules/600-flant-integration/templates/madison/deployment.yaml (hunk 6 above)
Comment (RomanenkoDenys) [reply ee/modules/600-flant-integration/templates/madison/deployment.yaml:59]: if necessary, we can add quotes to the helm lib function in the future.

File: testing/matrix/linter/rules/modules/images.go (hunk 8 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/modules/images.go:36]: yes, thanks !

File: testing/matrix/linter/rules/rules.go (hunk 10 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go:36]: ok, thanks, i change defaultRegistry to "registry.example.com/deckhouse"

File: testing/matrix/linter/rules/rules.go (hunk 11 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go:205]: ok, changed, thanks !

File: Makefile (hunk 18 above)
Comment (RomanenkoDenys) [reply Makefile:86]: Yes. I add comment to makefile

File: .github/workflows/build-and-test_release.yml (hunk 16 above)
Comment (RomanenkoDenys) [reply .github/workflows/build-and-test_release.yml:1465]: i do not understand what you mean. In matrix tests we use alpine go image, not builded deckhouse image.

File: testing/matrix/linter/rules/rules.go (hunk 7 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go:209]: Moved, thanks !

File: testing/matrix/linter/rules/rules.go (hunk 13 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go:217]: https://regex101.com/library/FKV8ot, added as comment

File: testing/matrix/linter/rules/rules.go (hunk 14 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go:212]: Yes, good point, thanks !

File: testing/matrix/linter/rules/rules.go (hunk 15 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go:93]: Thanks, this one from first version

File: .github/workflows/build-and-test_dev.yml (hunk 17 above)
Comment (RomanenkoDenys) [reply .github/workflows/build-and-test_dev.yml:934]: yes, we use make now, thanks !

File: testing/matrix/linter/rules/rules.go (hunk 12 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go]: i used go-containerregistry functions to parse image string, thanks !

File: testing/matrix/linter/rules/rules.go (hunk 9 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go:176]: Added comments

**Comment by RomanenkoDenys** [comment]
One problem is unsolved - when local matrix tests fail, make stops executing task, and symlinks cannot restore.

File: Makefile (hunk 18 above)
Comment (RomanenkoDenys) [reply Makefile]: as an option: after all the tests are run before the build, we can remove the symlinks, and add them at the build stage.

File: .github/workflows/build-and-test_release.yml (hunk 16 above)
Comment (nabokihms) [reply .github/workflows/build-and-test_release.yml:1465]: Previously, there was a variable. Now the image is hardcoded.

File: docs/documentation/pages/internal/DEVELOPMENT.md [hunk 19]
```diff
@@ -380,6 +380,35 @@ Usage:
 {{- end }}
 ```
 
+### Container images
```
Comment (nabokihms) [inline docs/documentation/pages/internal/DEVELOPMENT.md:383]: Does not seem related to this PR.

File: testing/matrix/linter/rules/rules.go [hunk 20]
```diff
@@ -166,24 +171,68 @@ func containerEnvVariablesDuplicates(object storage.StoreObject, containers []v1
...
+		if registry != defaultRegistry {
 			return errors.NewLintRuleError("CONTAINER003",
 				object.Identity()+"; container = "+c.Name,
 				nil,
-				"Image tag \"latest\" used",
+				"All images must be deployed from the same default registry - "+defaultRegistry,
 			)
 		}
+
+		switch os.Getenv("CI") {
```
Comment (nabokihms) [inline testing/matrix/linter/rules/rules.go:217]: Do we still need this branch if there is no dependency on the images_tags.json file?

File: Makefile [hunk 21]
```diff
@@ -79,10 +79,16 @@ deps: bin/golangci-lint bin/trivy bin/regcopy bin/jq bin/yq bin/crane ## Install
 tests-modules: ## Run unit tests for modules hooks and templates.
 	go test -timeout=${TESTS_TIMEOUT} -vet=off ./modules/... ./global-hooks/... ./ee/modules/... ./ee/fe/modules/...
 
+.ONESHELL:
 tests-matrix: ## Test how helm templates are rendered with different input values generated from values examples.
   ##~ Options: FOCUS=module-name
-	go test ./testing/matrix/ -v
-
+  # TODO: Need to find a proper way to maintain these symlinks.
```
Comment (nabokihms) [inline Makefile:85]: Thinking about it a little bit more.
1. Maybe it is better to move setup and teradown to the golang tests.
2. How do we test OpenStack and vSphere cloud providers?  I assume there are checksum helpers in these folders.

File: .github/workflows/build-and-test_release.yml [hunk 22]
```diff
@@ -1462,23 +1462,15 @@ jobs:
...
-          TESTS_IMAGE_NAME=$(echo ${TESTS_IMAGE_NAME} | base64 -d | gunzip)
-
           # Print image name in uppercase to prevent hiding non-secret registry host stored in secret.
           echo "Tests image name: '${TESTS_IMAGE_NAME}'" | tr '[:lower:]' '[:upper:]'
           echo "⚓️ 📥 [$(date -u)] Pull 'tests' image..."
           docker pull ${TESTS_IMAGE_NAME}
           echo "⚓️ 🏎 [$(date -u)] Run tests..."
-          docker run -w /deckhouse -v ${{env.HOME}}/go-pkg-cache:/go/pkg ${TESTS_IMAGE_NAME} go test -timeout=${{env.TEST_TIMEOUT}} ./testing/matrix/ -v
-    # </template: tests_template>
+          docker run -w /deckhouse -v ${{github.workspace}}:/deckhouse -e "TERM=xterm-256color" -e "CGO_ENABLED=0" -v ${{env.HOME}}/go-pkg-cache:/go/pkg ${TESTS_IMAGE_NAME} sh -c "apk add -f make; make tests-matrix"
```
Comment (nabokihms) [inline .github/workflows/build-and-test_release.yml:1472]: > apk add -f make

Oh no.

Comment (nabokihms) [inline .github/workflows/build-and-test_release.yml:1472]: TESTS_IMAGE_NAME is constant now, but it is still stored in the cache and prepulled, and printed.

File: ee/modules/600-flant-integration/templates/madison/deployment.yaml (hunk 6 above)
Comment (nabokihms) [reply ee/modules/600-flant-integration/templates/madison/deployment.yaml:59]: This is not possible, because it will break this helper usage in cases like this:
```
"--acme-http01-solver-image={{ include "helm_lib_module_image" (list . "certManagerAcmeSolver") }}"
```

File: .github/workflows/build-and-test_release.yml (hunk 22 above)
Comment (RomanenkoDenys) [reply .github/workflows/build-and-test_release.yml:1472]: Changed image to golang-debian, which have preinstalled make

File: Makefile (hunk 21 above)
Comment (RomanenkoDenys) [reply Makefile:85]: 1. Moved work with symlinks to matrix tests code.
2. This case is taken into consideration.

File: docs/documentation/pages/internal/DEVELOPMENT.md (hunk 19 above)
Comment (RomanenkoDenys) [reply docs/documentation/pages/internal/DEVELOPMENT.md:383]: This is because we now use helm image helpers in all controllers, so we need a devel doc for it.

**Review by kvaps — APPROVED** [review]
reviewed modules: linstor, local-path-provisioner, snapshot-controller
LGTM

File: ee/modules/600-flant-integration/templates/madison/deployment.yaml (hunk 6 above)
Comment (shvgn) [reply ee/modules/600-flant-integration/templates/madison/deployment.yaml:59]: Indeed, quoting will not solve the formatting problem. 
Anyway, it is not a blocker.

File: .github/workflows/build-and-test_release.yml (hunk 22 above)
Comment (RomanenkoDenys) [reply .github/workflows/build-and-test_release.yml]: Adds base_images to gomplate templates

File: testing/matrix/linter/rules/rules.go (hunk 20 above)
Comment (RomanenkoDenys) [reply testing/matrix/linter/rules/rules.go]: No, removed

File: .github/workflows/build-and-test_release.yml (hunk 16 above)
Comment (RomanenkoDenys) [reply .github/workflows/build-and-test_release.yml]: changed to render from image_versions.yaml

File: Makefile (hunk 18 above)
Comment (kvaps) [reply Makefile]: I just described my ideas in the https://github.com/deckhouse/deckhouse/issues/2217

File: testing/matrix/linter/run.go [hunk 23]
```diff
@@ -52,6 +131,12 @@ func Run(tmpDir string, m utils.Module) error {
 
 	var values []string
 	var err error
+
+	err = changeSymlinks(m)
```
Comment (nabokihms) [inline testing/matrix/linter/run.go:135]: For testing, it is better to use a proper way to setup and teardown a test suit. I previously proposed to use 
https://pkg.go.dev/github.com/stretchr/testify/suite#SetupTestSuite

It will help to restore symlinks in any case even if test failed to run.

File: testing/matrix/linter/run.go [hunk 24]
```diff
@@ -39,6 +39,85 @@ func isExist(baseDir, filename string) bool {
...
+// restoreSymlinks restores symlinks in module dir to original place
+func restoreSymlinks(m utils.Module) error {
+	switch m.Name {
+	case "cloud-provider-openstack":
+		err := changeSymlink("/deckhouse/ee/modules/030-cloud-provider-openstack/candi", "/deckhouse/candi/cloud-providers/openstack/")
+		if err != nil {
+			return err
+		}
+	case "cloud-provider-vsphere":
+		err := changeSymlink("/deckhouse/ee/modules/030-cloud-provider-vsphere/candi", "/deckhouse/candi/cloud-providers/vsphere/")
```
[the remark this answers is not in the document]
Comment (RomanenkoDenys) [reply testing/matrix/linter/run.go:100]: yes, good solution

File: testing/matrix/linter/run.go [hunk 25]
```diff
@@ -39,6 +39,85 @@ func isExist(baseDir, filename string) bool {
...
+			return err
+		}
+	}
+	return nil
+}
+
+// restoreSymlinks restores symlinks in module dir to original place
+func restoreSymlinks(m utils.Module) error {
+	switch m.Name {
+	case "cloud-provider-openstack":
```
[the remark this answers is not in the document]
Comment (RomanenkoDenys) [reply testing/matrix/linter/run.go:94]: yes

**Review by nabokihms — APPROVED** [review]
